폴더 구조 작업 및 API 1차 작업

GNOSss opened the pull request:
## #️⃣연관된 이슈

#2 

## 📝작업 내용

#2 외 

1. `controller`,`service`,`module`, `dto`, `entities` 폴더 생성하고
리소스 생성해서 각 폴더에 배치했습니다.

2. 화장실 상세정보 불러오는 API(`[GET] /detail/:id`) 작성했으나
테스트는 아직요

3. `toilet.entity.ts`에서 컬럼 철자가 틀려 수정
kids_toilet_femaie => kids_toilet_female


## 💬리뷰 요구사항(선택)
이후에 다시 요청하겠습니다.

GNOSss committed:
- commit message :
"""
rename: 리소스 생성 및 폴더 배치

`dto`,`entities`,`module`,`controller`,`service` 폴더 생성

`auth`,`likes`, `user.toilet.comment`, `detail.toilet` 리소스 생성 후 폴더에 배치

issue: feat/이슈번호
"""

GNOSss committed:
- commit message :
"""
feat: 화장실 상세정보

1차 초안

데이터베이스 세팅 후 테스트 예정

issue: [feat] 화장실 상세정보 제공 #2
"""

GNOSss committed:
- commit message :
"""
chore: 컬럼명 수정

kids_toilet_femaie => kids_toilet_female

issue: X
"""

GNOSss committed:
chore: Method 파라미터 누락

GNOSss committed:
chore: TypeOrm 각 테이블 세팅

추가적인 설명
확정된 세팅은 아닌 테스트로 사용중에 있음
추가로 수정 보완이 요구될 수 있음

issue: X

GNOSss committed:
- commit message :
"""
feat: 로그인 및 인증,
닉네임 설정,
액세스 토큰 연장,
로그아웃,
기능 구현

issue: [feat] 로그인 및 인증, 닉네임 설정, 로그아웃, 토큰 연장 #4
"""

GNOSss committed:
- commit message :
"""
chore: 화장실 테이블의 컬럼 옵션 변경

추가적인 설명
street_address , lot_address 컬럼의
속성으로 nullable: true로 설정함

"""

GNOSss committed:
- commit message :
"""
feat: 댓글 조회 기능 구현

추가적인 설명 :
비로그인 유저 `/comments/:toiletId/public`
로그인 유저 `/comments/:toiletId`

공통된 반환값에서
로그인한 상태에서는
`isMine`이라는 Boolean값을 보내줌으로써
프론트엔드에서 `''수정'',''삭제''` 버튼을 활성화 할 수 있습니다.

issue: [feat] 댓글 조회, 생성, 수정, 삭제 기능 구현  #6
"""

GNOSss committed:
fix: 데이터베이스 접근 오류 수정

GNOSss committed:
Merge branch 'main' of github.com:GeubDDong/Backend into dev

GNOSss committed:
Merge branch 'main' of github.com:GeubDDong/Backend into dev

yura0302 commented on src/controller/detail.toilet.controller.ts:
```diff
@@ -0,0 +1,12 @@
+  getDetailToilet(@Param() id: string) {
+    return this.detailToiletService.getDetailInfo(+id);
+  }
+}
```
이 부분에서 id가 string으로 들어오는데 이 부분을 +하면 NaN이 나올 것 같습니다 ! 혹시 콘솔 찍어보시거나 포스트맨으로 url 테스트해보신걸까요 ? 
제 생각에는 @Param("id") 이렇게 지정하면 아마 id값이 잘 들어 갈 것 같습니다 ! 확인 후 수정 부탁드려요 🙂

yura0302 commented on src/entities/toilet.entity.ts:
승수님 혹시 entities 부분만 복수로 설정하신 이유가 있으실까요 ? 제가 컨트롤러나 레포지토리는 단수형인데 여기만 복수형이라 혹시 다른 이유가 있어서 그러신가해서 여쭤봅니다 ! 보통은 컨벤션으로 단수를 많이 사용하는거같아서 여쭤봤어요 🙂

GNOSss replied:
entity로 바꿔 놔야겠어요 , 만들다 보니 통일성을 놓쳤네요 !

yura0302 commented on src/entities/toilet.entity.ts:
```diff
@@ -0,0 +1,55 @@
+  disabled_female: number;
+
+  @Column({ type: 'int' })
+  kids_toilet_femaie: number;
```
이 부분 femaie 오타 난 것 같은데 확인부탁드려요

GNOSss replied:
[feac50f](https://github.com/GeubDDong/Backend/pull/3/commits/feac50f1d1577673a72d37257eae6fa0da34d8d2)
에서 수정완료 시켰습니다 !

yura0302 commented on src/service/detail.toilet.service.ts:
```diff
@@ -0,0 +1,31 @@
+      disabled_male: toiletInfo.disabled_male > 0 ? 'Y' : 'N',
+      kids_toilet_male: toiletInfo.kids_toilet_male > 0 ? 'Y' : 'N',
+      disabled_female: toiletInfo.disabled_female > 0 ? 'Y' : 'N',
+      kids_toilet_female: toiletInfo.kids_toilet_femaie > 0 ? 'Y' : 'N',
```
이건 저도 찾아봐야 할 것 같은데 이 부분을 DTO에서 변환하게 되면 서비스코드가 깔끔해질것같은데 승수님 생각은 어떠세요 ?

GNOSss replied:
해당 DTO에서 생성자 만들어서 Obejct.assign 사용해서 처리했습니다. (안전 객체 복사)

yura0302 commented on src/service/detail.toilet.service.ts:
```diff
@@ -0,0 +1,31 @@
+    });
+
+    if (!toiletInfo) {
+      throw new NotFoundException('찾을 수 없는 id입니다.');
```
이 오류 메시지 부분을 혹시 
```
if (!toiletInfo) {
  throw new NotFoundException(`${id}에 해당하는 화장실 정보를 찾을 수 없습니다.`);
}
```
로 바꿔보는건 어떨까요 ? 그럼 더 명확할 것 같습니다. 🙂

GNOSss replied:
수정 했습니다~

yura0302 commented on src/controller/detail.toilet.controller.ts:
```diff
@@ -6,7 +6,7 @@ export class DetailToiletController {
 
   @Get(':id')
-  getDetailToilet(@Param() id: string) {
+  getDetailToilet(@Param('id') id: string) {
```
앗 제가 늦게 확인했네요 ㅎㅎ;

yura0302 commented on src/entities/likes.entity.ts:
```diff
@@ -0,0 +1,22 @@
+  @PrimaryColumn({ type: 'varchar', length: 200 })
+  user_email: string;
+
+  @PrimaryColumn()
```
이 @PrimaryColumn()는 필드가 기본키(Primary Key)역할을 한다는 의미인데 지금 이 toilet_id는 왜래키 같아보여요 만약, toilet_id가 외래키가 맞다면 @PrimaryColumn() 보다는 @Column()을 사용하시는게 어떨까요? 🙂

GNOSss replied:
user_email과 toilet_id가 복합키로 설정해보려고했습니다.
한 명의 유저(user_email)가 같은 화장실(toilet_id)에 여러 번 좋아요(Like)할 수 없도록 설정하려고 의도했는데
다른 방법이 있을지 몰라서 위 방법으로 적용해봤습니다 ㅎ;

yura0302 commented on src/entities/toilet.entity.ts:
```diff
@@ -52,4 +54,10 @@ export class ToiletModel {
+  @OneToMany(() => UserToiletCommentModel, (comment) => comment.toilet)
+  comments: UserToiletCommentModel[];
+
+  @OneToMany(() => LikesModel, (like) => like.toilet_id)
+  likes: LikesModel[];
```
like.toilet_id는 LikesModel의 필드이름인데 ManytoOne 관게예서 정의한 toilet 객체를 사용하는게 맞을거같아요 
likes 관계는 like.toilet을 참조해야 제대로 동작할 것 같습니다. 
```
@OneToMany(() => LikesModel, (like) => like.toilet)
likes: LikesModel[];
```
toilet_id가 아니라 toilet을 참조하도록 수정해야할 것 같아요

GNOSss replied:
수정 했습니다 !

yura0302 commented on src/entities/users.entity.ts:
```diff
@@ -0,0 +1,42 @@
+  @PrimaryColumn({ type: 'varchar', length: 200 })
+  email: string;
+
+  @Column({ type: 'bigint' })
+  id: number;
```
그리고 이건 저도 겪었던 문제인데 typeORM에서는 bigint를 사용할 때 number대신 string을 권장하고있습니다. 
해당 이슈 참고하시면 좋을 것 같아요 https://github.com/typeorm/typeorm/issues/2400

GNOSss replied:
확인했습니다. 코드 수정 들어가야될 거 같아요

yura0302 commented on src/controller/auth.controller.ts:
```diff
@@ -1,7 +1,91 @@
+  @Get('kakao-callback')
+  @UseGuards(AuthGuard('kakao'))
+  @HttpCode(301)
+  async kakaoLogin(@Req() req: Request, @Res() res: Response) {
+    const user = req.user;
+
+    const { accessToken, refreshToken } = await this.authService.getStoreTokens(
+      user.id,
+    );
+
+    res.cookie('refreshToken', refreshToken, { httpOnly: true, secure: false });
+
+    return res.status(user.isNewUser ? 201 : 200).json({
+      message: 'success',
+      accessToken,
+      statusCode: user.isNewUser ? 201 : 200,
+    });
+  }
```
@Res()를 사용하면 Nest의 표준 방식이 자동으로 비활성화 됩니다. 
그리고 nest에서는 @Res를 사용하지 않고 return 으로 응답을 처리하도록 권장합니다. 
또 return 할 때 refreshToken도 같이 넣으면 될 것 같아요 
ref. https://docs.nestjs.com/controllers#library-specific-approach

GNOSss replied:
보내준 URL 참고해서 
`CookieOptions` 와 `@Res({passthrough : true})` 적용해봤습니다.
```ts
async kakaoLogin(
    @Req() req: Request,
    @Res({ passthrough: true }) res: Response,
  ) {
    const user = req.user;

    const { accessToken, refreshToken } = await this.authService.getStoreTokens(
      user.id,
    );

    const cookieOptions: CookieOptions = { httpOnly: true, secure: false };
    res.cookie('RefreshToken', refreshToken, cookieOptions);

    return {
      message: 'success',
      accessToken,
      statusCode: user.isNewUser ? 201 : 200,
    };
  }
  ```

yura0302 commented on src/controller/auth.controller.ts:
```diff
@@ -1,7 +1,91 @@
+  @Public()
+  @UseGuards(AuthGuard('kakao'))
+  @Get('login/:provider')
+  checkProvider() {}
```
이건 어떤 기능을 하는 메서드 인가요 ? 
유효성 검사를 수행하시려면 @Param안에 provider가 string인지 받는게 나을 것 같습니다

GNOSss replied:
현재는 카카오 소셜 로그인만 적용 중이라서
`checkProvider` 함수에 다른 로직을 넣어 두지 않았습니다.

만약 네이버, 구글 소셜 로그인을 적용하게되면
```ts
@Public()
@UseGuards(AuthGuard('kakao'), AuthGuard('google'), AuthGuard('naver'))
@Get('login/:provider')
checkProvider(@Param('provider') provider: string) {
  if (!['kakao', 'google', 'naver'].includes(provider)) {
    throw new BadRequestException('지원하지 않는 로그인 제공자입니다.');
  }
}
```
유효성 검사 로직을 넣어두려고 했습니다

yura0302 commented on src/controller/auth.controller.ts:
```diff
@@ -1,7 +1,91 @@
+    );
+    return res.json({ statusCode, message });
+  }
 }
```
req.body를 직접 사용하시는 것 보다 DTO를 사용하시는건 어떠세요 ? 
class-validator를 이용해서 문자열인지, 길이는 몇인지를 수행하는게 좋아보일 듯 합니다.

GNOSss replied:
코드 작성하고나서 유효성 검사를 안넣어서 고려했었습니다 ^^;;

main.ts 
```ts
app.useGlobalPipes(new ValidationPipe());
```

set.nickname.dto.ts
```ts
import { Transform } from 'class-transformer';
import { IsString, Matches, MaxLength } from 'class-validator';

export class SetNicknameDto {
  @Transform(({ value }) => String(value))
  @IsString()
  @MaxLength(10, { message: '닉네임은 최대 10글자까지만 가능합니다.' })
  @Matches(/^[가-힣a-zA-Z][가-힣a-zA-Z0-9]*$/, {
    message:
      '닉네임은 한글, 영문, 숫자만 사용 가능하며 첫 글자는 숫자가 될 수 없습니다.',
  })
  nickname: string;
}

```

user.entity.ts
```ts
  @Post('nickname')
  @UsePipes(new ValidationPipe({ transform: true }))
  async setNickname(
    @Req() req: Request,
    @Body() setNicknameDto: SetNicknameDto,
  ) {
    const userId = req.user.id;
    const { nickname } = setNicknameDto;

    const { statusCode, message } = await this.authService.setNickname(
      userId,
      nickname,
    );
    return { statusCode, message };
  }
```

이렇게 진행해봤습니다.

yura0302 reviewed (approved):
수고하셨습니다 🙂

GNOSss committed:
- commit message :
"""
rename : Entities -> Entity 폴더명 변경
issue: [rename] Modify from 'Entities' Folder to 'Entity' Folder #8

---

refactor: id 컬럼 bigint -> string으로 변경
issue: [refactor] from ({ type : 'bigint'}) to ({ type : 'string' }) #12

---

refactor: from 'like.toilet_id' to 'like.toilet'

추가적인 설명 :
중간에 commit을 못해서
수정된 부분이 마킹되지 않았습니다.
(17line)이 수정된 부분입니다.

"""

GNOSss committed:
- commit message :
"""
refactor: if(toiletInfo) 오류 구문 수정
issue: [refactor] if(!toiletInfo) 오류 구문 수정 #10

---

feractor: detailToiletResponseDTO 생성

추가적인 설명:
해당 DTO 에서 constructor 생성 후
일부 컬럼들의 조건문 처리 후 반환

issue: [refactor] add validate in DetailToiletResponseDto  #9

"""

GNOSss committed:
- commit message :
"""
refactor: add setNickname DTO

추가적인 설명
auth.controller.ts에 적용 완료했습니다.

issue: [refactor] add setNickname DTO #14
"""

GNOSss committed:
- commit message :
"""
feat: 좋아요 기능 구현
issue: [feat] 좋아요 조회, 추가, 삭제 기능 구현 #7
"""

GNOSss committed:
- commit message :
"""
feat: return문을 Nest표준 방식으로 코드 수정

issue: [refactor] return문을 Nest 표준 방식으로 코드 수정 #13
"""

GNOSss committed:
- commit message :
"""
feat: 댓글 수정, 삭제 기능 구현

issue: [feat] 댓글 조회, 생성, 수정, 삭제 기능 구현 #6
"""

GNOSss committed:
- commit message :
"""
chore: 기타 수정 사안

추가적인 설명
- 데이터 타입 변환
- 불필요한 파일 호출 제거 및 추가
- return문 형식 변경
등의 수정안입니다.

"""

GNOSss committed:
- commit message :
"""
refactor: toilet엔터티 default 값 추가

추가적인 설명
default : '정보없음' 으로 추가함

"""
